cli/src/check.rs

```diff
@@ -0,0 +1,197 @@
+use anyhow::Result;
+use k8s_openapi::{
+    api::{apps::v1::Deployment, core::v1::Namespace},
+    apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition,
+};
+use kube::{api::ListParams, Api, Client, ResourceExt};
+use serde::Serialize;
+
+const CHECK: &str = "√";
+const EX: &str = "×";
+
+#[derive(Serialize)]
+struct CheckOutput {
+    success: bool,
+    categories: Vec<Category>,
+}
+
+#[derive(Serialize)]
+struct Category {
+    category_name: &'static str,
+    checks: Vec<CheckResult>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "lowercase")]
+enum CheckStatus {
+    Success,
+    Error,
+}
+
+#[derive(Serialize, Default)]
+pub struct CheckResult {
+    description: &'static str,
+    result: CheckStatus,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    hint: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    error: Option<String>,
+}
+
+impl Default for CheckStatus {
+    fn default() -> CheckStatus {
+        CheckStatus::Success
+    }
+}
+
+impl CheckResult {
+    fn new(description: &'static str) -> CheckResult {
+        CheckResult {
+            description,
+            ..Default::default()
+        }
+    }
+}
+
+pub async fn traffic_split_check(client: Client) -> CheckResult {
+    let api = Api::<CustomResourceDefinition>::all(client.clone());
+    let mut result = CheckResult::new("TrafficSplit CRD exists");
+    match api.get_opt("trafficsplits.split.smi-spec.io").await {
+        Result::Ok(Some(_)) => result.result = CheckStatus::Success,
+        Result::Ok(None) => {
+            result.result = CheckStatus::Error;
+            result.error = Some("TrafficSplit CRD is not installed".into());
+            result.hint = Some("https://github.com/linkerd/linkerd-smi");
+        }
+        Result::Err(err) => {
+            result.result = CheckStatus::Error;
+            result.error = Some(err.to_string());
+            result.hint = Some("https://github.com/linkerd/linkerd-failover#troubleshooting");
+        }
+    };
+    result
```

Comment:
nit, take it or leave it: it might be a little nicer to just change the `match` here to return the `CheckStatus` rather than mutating it in place, like this:
```suggestion
    let result = CheckStatus::new("TrafficSplit CRD exists");
    match api.get_opt("trafficsplits.split.smi-spec.io").await {
        Ok(Some(_)) => result,
        Ok(None) => CheckStatus {
            result: CheckStatus::Error,
            error: Some("TrafficSplit CRD is not installed".into()),
            hint: Some("https://github.com/linkerd/linkerd-smi"),
            ..result
        }
        Err(err) => CheckStatus {
            result: CheckStatus::Error,
            error: Some(err.to_string()),
            hint: Some("https://github.com/linkerd/linkerd-failover#troubleshooting"),
            ..result
        }
    }
```